# Segmenting and Clustering Neighborhoods in Toronto

## Applied Data Science Capstone - Week 3

### 1. Data Pre-Processing

In [1]:
# Importing some libraries
import numpy as np
import pandas as pd
import requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [2]:
# Wikipedia page's url
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [3]:
# Getting the first table from Wikipedia page
df = pd.read_html(url)[0]

In [4]:
# Checking if df is correct
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Postal Code    180 non-null    object
 1   Borough        180 non-null    object
 2   Neighbourhood  180 non-null    object
dtypes: object(3)
memory usage: 4.3+ KB


In [6]:
# Dropping those rows whose Borough is "Not assigned"
df = df[df['Borough'] != 'Not assigned']

In [7]:
# Checking df once more
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 2 to 178
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Postal Code    103 non-null    object
 1   Borough        103 non-null    object
 2   Neighbourhood  103 non-null    object
dtypes: object(3)
memory usage: 3.2+ KB


In [9]:
df['Postal Code'].unique().size

103

There are 103 entries, and also 103 unique Postal Code. Therefore, every postal code indicates its neighborhoods in one row. There is no need of further pre-processing.

In [10]:
# Ckecking for 'Not assigned' Neighborhoods
df[df['Neighbourhood'] =='Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

There is no 'Not assigned' Neighborhood in df.

In [11]:
df.shape

(103, 3)

### 2. Getting Neighborhoods Coordinates

In [12]:
# Using CSV file provided by Coursera
coord = pd.read_csv('Geospatial_Coordinates.csv')

In [13]:
coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [14]:
# Merging df and coord dataframes on 'Postal Code' label
result = pd.merge(df,coord,on='Postal Code')

In [15]:
result

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [28]:
result_toronto = result[result['Borough'].str.contains('Toronto')]

In [32]:
result_toronto

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                         Neighbourhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                   North Toronto West, Lawrence Park  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

### 3. Clustering Toronto Neighborhoods

#### 3.1 Showing Postal Code on Toronto's map

In [33]:
from geopy.geocoders import Nominatim

from sklearn.cluster import KMeans

import folium

In [34]:
address = 'Toronto, On'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [37]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add Postal Code markers to map
for pc, lat, lng, borough, neighborhood in zip(result['Postal Code'], 
                                               result['Latitude'], result['Longitude'], result['Borough'], result['Neighbourhood']):
    label = f'{pc}, {neighborhood}, {borough}'
    label = folium.Popup(label, parse_html=True)
    folium.vector_layers.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
#map_newyork

#### 3.2 Organizing for Clustering

Foursquare Credentials and Version

In [38]:
CLIENT_ID = 'UVQYTO1U3J0FFQAHZIVP2ZXP5M3EYSCVPOMOEXYQ0DL1XDRK' # your Foursquare ID
CLIENT_SECRET = 'AQL0YLQ3WQAHQPIDNWXPC1EEJ5QW5C0Z3ZQG4U4DGRIASYZY' # your Foursquare Secret
VERSION = '20180605'
LIMIT = 200
radius = 500

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UVQYTO1U3J0FFQAHZIVP2ZXP5M3EYSCVPOMOEXYQ0DL1XDRK
CLIENT_SECRET:AQL0YLQ3WQAHQPIDNWXPC1EEJ5QW5C0Z3ZQG4U4DGRIASYZY


The function below was copied form Manhatan example.

In [39]:
def getNearbyVenues(postalcodes, names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for postalcode, name, lat, lng in zip(postalcodes, names, latitudes, longitudes):
        # print(postalcode, ' ', name)    
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            postalcode,
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code',
                             'Neighborhood', 
                             'Neighborhood Latitude',
                             'Neighborhood Longitude',
                             'Venue',
                             'Venue Latitude',
                             'Venue Longitude',
                             'Venue Category']
    
    return(nearby_venues)

Applying the function above to every Postal Code

In [40]:
toronto_venues = getNearbyVenues(postalcodes = result['Postal Code'],
                                 names=result['Borough'],
                                 latitudes=result['Latitude'],
                                 longitudes=result['Longitude']
                                  )

In [41]:
toronto_venues.size

17296

In [42]:
toronto_venues.head()

Postal Code Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         M3A   North York              43.753259              -79.329656   
1         M3A   North York              43.753259              -79.329656   
2         M3A   North York              43.753259              -79.329656   
3         M4A   North York              43.725882              -79.315572   
4         M4A   North York              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                Sun Life       43.754760       -79.332783   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

               Venue Category  
0                        Park  
1  Construction & Landscaping  
2           Food & Drink Shop  
3                Hockey Arena  
4       Portuguese Restaurant

In [43]:
toronto_venues.shape

(2162, 8)

#### Let's see how many venues there are in each Postal Code

In [44]:
toronto_venues.groupby('Postal Code').count()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
Postal Code                                                                
M1B                     1                      1                       1   
M1C                     3                      3                       3   
M1E                     8                      8                       8   
M1G                     3                      3                       3   
M1H                     9                      9                       9   
...                   ...                    ...                     ...   
M9N                     2                      2                       2   
M9P                     7                      7                       7   
M9R                     4                      4                       4   
M9V                     9                      9                       9   
M9W                     2                      2                       2   

             Venue  Venue Latitude  Venue Longitude  Venue Category  
Postal Code                                                          
M1B              1               1                1               1  
M1C              3               3                3               3  
M1E              8               8                8               8  
M1G              3               3                3               3  
M1H              9               9                9               9  
...            ...             ...              ...             ...  
M9N              2               2                2               2  
M9P              7               7                7               7  
M9R              4               4                4               4  
M9V              9               9                9               9  
M9W              2               2                2               2  

[100 rows x 7 columns]

It is necessary to get_dummies form 'Venue Category' column. 

In [68]:
#one hot encoding
toronto_venues_categories = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

In [69]:
toronto_venues_categories.shape

(2162, 272)

In [70]:
toronto_venues_categories_merged = toronto_venues.merge(toronto_venues_categories,on=toronto_venues.index)

In [71]:
toronto_venues_categories_merged.drop(['key_0'],axis=1,inplace=True)

In [72]:
toronto_venues_categories_merged.shape

(2162, 280)

Lets group rows by Postal Code, taking the total amount of a given venue

In [81]:
toronto_grouped = toronto_venues_categories_merged.groupby('Postal Code').mean().reset_index()
toronto_grouped.head()

Postal Code  Neighborhood Latitude  Neighborhood Longitude  Venue Latitude  \
0         M1B              43.806686              -79.194353       43.807448   
1         M1C              43.784535              -79.160497       43.783618   
2         M1E              43.763573              -79.188711       43.766320   
3         M1G              43.770992              -79.216917       43.770559   
4         M1H              43.773136              -79.239476       43.774519   

   Venue Longitude  Accessories Store  Afghan Restaurant  Airport  \
0       -79.199056                0.0                0.0      0.0   
1       -79.160541                0.0                0.0      0.0   
2       -79.191291                0.0                0.0      0.0   
3       -79.219579                0.0                0.0      0.0   
4       -79.240678                0.0                0.0      0.0   

   Airport Food Court  Airport Lounge  ...  Train Station  \
0                 0.0             0.0  ...            0.0   
1                 0.0             0.0  ...            0.0   
2                 0.0             0.0  ...            0.0   
3                 0.0             0.0  ...            0.0   
4                 0.0             0.0  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 277 columns]

In [82]:
toronto_grouped.shape

(100, 277)

#### 3.3 Clustering

In [83]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop(['Postal Code','Neighborhood Latitude', 'Neighborhood Longitude', 'Venue Latitude', 'Venue Longitude'], axis=1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([5, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [84]:
toronto_grouped = toronto_grouped.join(pd.DataFrame(kmeans.labels_,columns=['Cluster Label']))

In [85]:
toronto_grouped.head()

Postal Code  Neighborhood Latitude  Neighborhood Longitude  Venue Latitude  \
0         M1B              43.806686              -79.194353       43.807448   
1         M1C              43.784535              -79.160497       43.783618   
2         M1E              43.763573              -79.188711       43.766320   
3         M1G              43.770992              -79.216917       43.770559   
4         M1H              43.773136              -79.239476       43.774519   

   Venue Longitude  Accessories Store  Afghan Restaurant  Airport  \
0       -79.199056                0.0                0.0      0.0   
1       -79.160541                0.0                0.0      0.0   
2       -79.191291                0.0                0.0      0.0   
3       -79.219579                0.0                0.0      0.0   
4       -79.240678                0.0                0.0      0.0   

   Airport Food Court  Airport Lounge  ...  Vegetarian / Vegan Restaurant  \
0                 0.0             0.0  ...                            0.0   
1                 0.0             0.0  ...                            0.0   
2                 0.0             0.0  ...                            0.0   
3                 0.0             0.0  ...                            0.0   
4                 0.0             0.0  ...                            0.0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0               0.0          0.0                    0.0              0.0   
1               0.0          0.0                    0.0              0.0   
2               0.0          0.0                    0.0              0.0   
3               0.0          0.0                    0.0              0.0   
4               0.0          0.0                    0.0              0.0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  Cluster Label  
0       0.0          0.0            0.0          0.0              5  
1       0.0          0.0            0.0          0.0              0  
2       0.0          0.0            0.0          0.0              0  
3       0.0          0.0            0.0          0.0              0  
4       0.0          0.0            0.0          0.0              0  

[5 rows x 278 columns]

In [86]:
toronto_grouped.shape

(100, 278)

In [87]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_grouped['Neighborhood Latitude'], toronto_grouped['Neighborhood Longitude'], toronto_grouped['Postal Code'], toronto_grouped['Cluster Label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.vector_layers.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters